### PR TITLE
Globals

### PR DESCRIPTION
**Pull request by turian**
We not just pass around SynthGlobals everywhere. This makes our lives easier, as you will see in the next PR. Merge #111 first

### CODE REVIEW

File: torchsynth/module.py [hunk 1]
```diff
@@ -379,12 +337,20 @@ class TorchVCO(TorchSynthModule1D):
...
+    def __init__(
+        self,
+        midi_f0: T,
+        mod_depth: T,
+        synthglobals: TorchSynthGlobals,
+        phase: Optional[T] = None,
+    ):
+        super().__init__(synthglobals)
+        print("synthglobals", self.synthglobals)
+        print(self.batch_size)
```
Comment (jorshi) [inline torchsynth/module.py:353]: rogue prints?

File: torchsynth/module.py [hunk 2]
```diff
@@ -470,17 +436,19 @@ class TorchSineVCO(TorchVCO):
     midi_f0 (T)     :   pitch value in 'midi' (69 = 440Hz).
     mod_depth (T)   :   depth of the pitch modulation in semitones.
     phase (T)       :   initial phase values
-    **kwargs        :   keyword args, see TorchVCO
+    synthglobals: TorchSynthGlobals        :   keyword args, see TorchVCO
```
Comment (jorshi) [inline torchsynth/module.py:439]: update description: global args, see TorchSynthModule

File: torchsynth/module.py [hunk 3]
```diff
@@ -502,11 +470,19 @@ class TorchFmVCO(TorchVCO):
     midi_f0 (T)     :   pitch value in 'midi' (69 = 440Hz).
     mod_depth (T)   :   depth of the frequency (0-127)
     phase (T)       :   initial phase values
-    **kwargs        :   keyword args, see TorchVCO
+    synthglobals: TorchSynthGlobals        :   keyword args, see TorchVCO
```
Comment (jorshi) [inline torchsynth/module.py:473]: same -- update description

File: torchsynth/module.py [hunk 4]
```diff
@@ -536,18 +512,20 @@ class TorchSquareSawVCO(TorchVCO):
     mod_depth (T)   :   depth of the pitch modulation in semitones.
     shape (T)       :   Waveshape - square to saw [0,1]
     phase (T)       :   initial phase values
-    **kwargs        :   keyword args, see TorchVCO
+    synthglobals: TorchSynthGlobals        :   keyword args, see TorchVCO
```
Comment (jorshi) [inline torchsynth/module.py:515]: here too

File: torchsynth/module.py [hunk 5]
```diff
@@ -577,17 +555,17 @@ def partials_constant(self):
...
 
-class TorchVCA(TorchSynthModule1D):
+class TorchVCA(TorchSynthModule):
     """
     Voltage controlled amplifier.
 
     Parameters
     ----------
-    **kwargs : keyword args, see TorchSynthModule1D
+    synthglobals: TorchSynthGlobals : keyword args, see TorchSynthModule
```
Comment (jorshi) [inline torchsynth/module.py:564]: global args

File: torchsynth/module.py [hunk 6]
```diff
@@ -631,12 +609,109 @@ def noise_of_length(audio_in: Signal) -> Signal:
...
+        self.torchparameters[parameter_id].data = T(value)
+
+    def p(self, parameter_id: str) -> T:
+        """
+        Convenience method for getting the parameter value.
+        """
+        return self.torchparameters[parameter_id].from_0to1()
+
+
+class TorchSynthParameters(TorchSynthModule0Ddeprecated):
```
Comment (jorshi) [inline torchsynth/module.py:705]: Can this just derive from TorchSynthModule?

File: torchsynth/module.py [hunk 7]
```diff
@@ -654,27 +729,46 @@ class TorchSynth(nn.Module):
...
+        self,
+        synthglobals: TorchSynthGlobals,
+    ):
         super().__init__()
-        self.sample_rate = T(sample_rate)
-        self.buffer_size = T(buffer_size)
+        self.synthglobals = synthglobals
 
-        # Global Parameter Module
-        self.global_params = TorchSynthParameters(sample_rate, buffer_size)
```
Comment (jorshi) [inline torchsynth/module.py:663]: so this parameter is for stuff that is not within children SynthModules -- like master volume or something. Not quite the same as the new synth globals. But maybe you address this in the next PR. Will check out.

**Review by jorshi — COMMENTED** [review]
Love the TorchSynthGlobals idea. Super clean.

File: torchsynth/module.py (hunk 1 above)
Comment (turian) [reply torchsynth/module.py:353]: oups from debugging

File: torchsynth/module.py (hunk 2 above)
Comment (turian) [reply torchsynth/module.py:439]: fixed will push

File: torchsynth/module.py (hunk 7 above)
Comment (turian) [reply torchsynth/module.py:663]: yeah hope you like it

**Pull request merged**